### PR TITLE
RFR: Implemented a basic registration and login system

### PR DESCRIPTION
**Pull request by micleo2**
## Status: 
:rocket: Ready

## Description
Implemented the registration and login endpoints. Upon success, they return an authentication token which is valid for 24 hours. Required fields for the User table so far are: name, email, and password.

Related issues: #3

### CODE REVIEW

File: backend/Pipfile [hunk 1]
```diff
@@ -18,6 +18,8 @@ Flask-Cors = "*"
 flask-script = "*"
 sqlalchemy-utils = "*"
 flask = "*"
+pyjwt = "==1.4.2"
+bcrypt = "*"
```
Comment (XuanLi-Leong) [inline backend/Pipfile]: add specific version

File: backend/api/views/authenticate.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            message="Missing required information.", data={"status": "fail"}, status=422
+        )
+
+    user = User.query.filter_by(email=request.form["email"]).first()
+
+    if user is not None:
+        if bcrypt.checkpw(
+            request.form["password"].encode("utf8"), user.password.encode("utf8")
+        ):
+            auth_token = user.encode_auth_token()
```
Comment (XuanLi-Leong) [inline backend/api/views/authenticate.py]: try catch

File: backend/api/models/User.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                + datetime.timedelta(days=1, seconds=0),
+                "iat": datetime.datetime.utcnow(),
+                "sub": self.id,
+                "test": "working",
+            }
+            return jwt.encode(
+                payload, current_app.config.get("SECRET_KEY"), algorithm="HS256"
+            )
+        except Exception as e:
+            return e
```
Comment (XuanLi-Leong) [inline backend/api/models/User.py]: rethrow (or remove catching)

File: backend/api/views/register.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from flask import Flask, jsonify, request, Blueprint
```
Comment (XuanLi-Leong) [inline backend/api/views/register.py]: delete file

File: backend/Pipfile [hunk 5]
```diff
@@ -9,21 +9,23 @@ Flask-Migrate = "==2.1.1"
...
+"psycopg2" = "==2.7.5"
 pyflakes = "==1.6.0"
 requests = "==2.18.4"
 SQLAlchemy = "==1.1.14"
 "urllib3" = "==1.22"
-Flask-Cors = "*"
-flask-script = "*"
-sqlalchemy-utils = "*"
-flask = "*"
+Flask-Cors = "==3.0.6"
```
Comment (Davidw1339) [inline backend/Pipfile]: great job pegging :nail_care:

File: backend/tests/test_basic.py [hunk 6]
```diff
@@ -1,26 +1,59 @@
...
 
-    assert rs.status_code == 200
-    ret_dict = rs.json  # gives you a dictionary
-    assert ret_dict["success"] == True
-    assert ret_dict["result"]["persons"] == []
 
-    # create Person and test whether it returns a person
-    temp_person = Person(name="Tim")
-    db.session.add(temp_person)
+def test_double_register(client):
```
Comment (Davidw1339) [inline backend/tests/test_basic.py]: good test coverage :black_flag: :1st_place_medal:

File: backend/Pipfile (hunk 5 above)
Comment (XuanLi-Leong) [reply backend/Pipfile]: @hungryflamingcat

**Pull request merged**